### PR TITLE
SourceData Inspector

### PR DESCRIPTION
**Pull request by johannes-wolf**
## Changes
- Added a new common Inspector panel that contains all Feature & SourceData inspectors
- Moved the old inspector panel to FeatureInspector
- Fetching data now forwards the future so you can pipe it through `then` and friends
- Added a new ng directive to tailor PrimeNGs table view filtering to our needs
- Added a new ng pipe to highlight text
- Upgraded Angular to version 18.*

## UI Changes
- The scrollbar of tree views is now on the right
- We now use PrimeNGs virtual scrolling
- Upgraded PrimeNG to a patched 17.18.19, that includes our fixes for virtual scrolling

## Yet Missing & Bugs
- ~~The inspector title does not show the current feature~~
- ~~Inspecting source data does not get reflected in the windows URL~~
- PrimeNGs table scrolling is broken and needs some workaround to not randomly hide rows
  - Resizing the browser window retriggers the row calculation, maybe we can do something to trigger it after setting the table data
  - Sometimes it calculates a wrong number of items, so rows become invisible
  - Link to our PrimeNG PR https://github.com/primefaces/primeng/pull/16273
- ~~Setting a column size for tree tables does not work at all, horizontal scrolling also does not work~~ (won't fix)
- ~~Listening for feature selection changes should (maybe) be moved out of the FeatureInspector~~ (should be done separately)

### CODE REVIEW

File: libs/core/src/parser.cpp [hunk 1]
```diff
@@ -105,7 +110,20 @@ mapget::TileFeatureLayer::Ptr TileLayerParser::readTileFeatureLayer(const Shared
         {
             return resolveMapLayerInfo(std::string(mapId), std::string(layerId));
         },
-        [this](auto&& nodeId) { return cachedFieldDicts_->getFieldDict(nodeId); });
+        [this](auto&& nodeId) { return cachedStrings_->getStringPool(nodeId); });
+    return result;
+}
+mapget::TileSourceDataLayer::Ptr TileLayerParser::readTileSourceDataLayer(SharedUint8Array const& buffer)
```
Comment (josephbirkner) [inline libs/core/src/parser.cpp:116]: Missing newline :)

**Review by josephbirkner — APPROVED** [review]
Awesome changes, added some suggestions :+1:

File: erdblick_app/app/sourcedata.panel.component.ts [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+        let select = (node: TreeTableNode, parents: TreeTableNode[], highlight: boolean, virtualRowIndex: number) => {
+            if (!node.data) {
+                return;
+            }
+
+            if (highlight) {
+                node.data.styleClass = "highlight";
+            }
+
+            const address = node.data.address;
```
Comment (Waguramu) [inline erdblick_app/app/sourcedata.panel.component.ts:218]: Don't need to shadow the variable. Just use `node.data.address`

File: erdblick_app/app/sourcedata.panel.component.ts [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+
+        if (typeof address === 'object') {
+            return `${address.offset}:${address.size}`
+        } else {
+            return `${address}`
+        }
+    }
+
+    selectItemWithAddress(address: bigint) {
+        let searchAddress: any = address;
```
Comment (Waguramu) [inline erdblick_app/app/sourcedata.panel.component.ts:186]: No need to assign `searchAddress` here.

File: erdblick_app/app/sourcedata.panel.component.ts [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+
+            const addressLow = typeof searchAddress === 'object' ? searchAddress['offset'] : searchAddress;
+            const addressHigh = addressLow + (typeof searchAddress === 'object' ? searchAddress['size'] : searchAddress);
+
+            addressInRange = (addr: any) => {
+                return addr.offset >= addressLow && addr.offset + addr.size <= addressHigh && (addr.size != 0 || addressLow == addressHigh);
+            }
+        } else {
+            addressInRange = (addr: any) => {
+                return addr == searchAddress;
```
Comment (Waguramu) [inline erdblick_app/app/sourcedata.panel.component.ts:202]: Use `address`

File: erdblick_app/app/sourcedata.panel.component.ts [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+        }
+
+        const prefix = "https://developer.nds.live/schema/";
+
+        let match = schema.match(/^nds\.(([^.]+\.)+)v(\d{4}_\d{2})((\.[^.]*)+)/);
+        if (!match || match.length <= 4)
+            return schema;
+
+        // Sub-namespaces in front of the version get joined by "-". Names past the version get joined by "/"
+        let url =
```
Comment (Waguramu) [inline erdblick_app/app/sourcedata.panel.component.ts:166]: Very nice :star_struck: (but make it `const`)

File: erdblick_app/app/sourcedata.panel.component.ts [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+     * @return string HTML
+     */
+    schemaTypeURLFormatter(schema?: string) {
+        if (!schema) {
+            return schema;
+        }
+
+        const prefix = "https://developer.nds.live/schema/";
+
+        let match = schema.match(/^nds\.(([^.]+\.)+)v(\d{4}_\d{2})((\.[^.]*)+)/);
```
Comment (Waguramu) [inline erdblick_app/app/sourcedata.panel.component.ts:161]: `const match`

File: erdblick_app/app/sourcedata.panel.component.ts [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+     * Unsets the tree to an empty array.
+     *
+     * @param message Error message
+     */
+    setError(message: string) {
+        this.loading = false;
+        this.treeData = [];
+        this.errorMessage = message;
+
+        console.error(this.errorMessage);
```
Comment (Waguramu) [inline erdblick_app/app/sourcedata.panel.component.ts:144]: Add a description. Something like `console.error("Error while processing SourceData tree:", this.errorMessage)`

File: erdblick_app/app/parameters.service.ts [hunk 8]
```diff
@@ -159,7 +167,7 @@ export class ParametersService {
             }
         });
 
-    constructor(public router: Router) {
+    constructor(public router: Router /*, private inspectionService: InspectionService*/) {
```
Comment (Waguramu) [inline erdblick_app/app/parameters.service.ts:170]: To be removed?

File: erdblick_app/app/inspection.service.ts [hunk 9]
```diff
@@ -141,5 +171,48 @@ export class InspectionService {
...
+    async loadSourceDataLayer(tileId: number, layerId: string, mapId: string) : Promise<TileSourceDataLayer> {
+        console.log(`Loading SourceDataLayer layerId=${layerId} tileId=${tileId}`);
+
+        let requests = [{
+            mapId: mapId,
+            layerId: layerId,
+            tileIds: [tileId]
+        }];
+
+        let tileParser = new coreLib.TileLayerParser();
```
Comment (Waguramu) [inline erdblick_app/app/inspection.service.ts:183]: `const tileParser`

File: erdblick_app/app/inspection.service.ts [hunk 10]
```diff
@@ -141,5 +171,48 @@ export class InspectionService {
         }
     }
 
+    async loadSourceDataLayer(tileId: number, layerId: string, mapId: string) : Promise<TileSourceDataLayer> {
+        console.log(`Loading SourceDataLayer layerId=${layerId} tileId=${tileId}`);
+
+        let requests = [{
```
Comment (Waguramu) [inline erdblick_app/app/inspection.service.ts:177]: Could be inlined?

File: erdblick_app/app/inspection.service.ts [hunk 11]
```diff
@@ -141,5 +171,48 @@ export class InspectionService {
...
+            mapId: mapId,
+            layerId: layerId,
+            tileIds: [tileId]
+        }];
+
+        let tileParser = new coreLib.TileLayerParser();
+
+        let newRequestBody = JSON.stringify({
```
Comment (Waguramu) [inline erdblick_app/app/inspection.service.ts:185]: `const newRequestBody`

File: erdblick_app/app/inspection.service.ts [hunk 12]
```diff
@@ -141,5 +171,48 @@ export class InspectionService {
...
+                    }, blob);
+                } else {
+                    throw new Error(`Unknown message type ${messageType}.`)
+                }
+            });
+
+        return fetch.go()
+            .then(_ => {
+                if (!layer)
+                    throw new Error(`Error loading layer.`);
```
Comment (Waguramu) [inline erdblick_app/app/inspection.service.ts:212]: Could add more information to the error message. Maybe IDs?

File: erdblick_app/app/inspection.service.ts [hunk 13]
```diff
@@ -16,9 +17,24 @@ interface InspectionModelData {
...
+    tileId: number,
+    layerId: string,
+    address: bigint,
+    featureId: string,
+}
+
+export function selectedSourceDataEqualTo(a: SelectedSourceData | null, b: SelectedSourceData | null) {
+    if (!a || !b)
+        return false;
+    return (a == b || (a.mapId == b.mapId && a.tileId == b.tileId && a.layerId == b.layerId && a.address == b.address && a.featureId == b.featureId));
```
Comment (Waguramu) [inline erdblick_app/app/inspection.service.ts:35]: `a === b`

File: erdblick_app/app/inspection.panel.component.ts [hunk 14]
```diff
@@ -1,350 +1,153 @@
...
-            cursor: pointer;
-            text-decoration: underline dotted;
-            font-style: italic;
-        }
-        
-        @media only screen and (max-width: 56em) {
-            .resizable-container-expanded {
-                height: calc(100vh - 3em);;
+    styles: [
+        `@layer erdblick {
```
Comment (Waguramu) [inline erdblick_app/app/inspection.panel.component.ts:42]: Do we necessarily need layers?

**Pull request merged**